### ISSUE TITLE
Proposal: Improve `SignInAsync`

### ISSUE THREAD

**Issue by Jehoel** [IdentityServer/IdentityServer4#2650]
**Is your feature request related to a problem? Please describe.**

Currently IdentityServer4 requires applications use `AuthenticationManagerExtensions::SignInAsync` in order to complete the user authentication process when using grants and flows that use a browser-based login form presented using ASP.NET Core MVC.

This is a problem because ASP.NET Core comes with its own methods also named `SignInAsync` and to a software developer new to IdentityServer (e.g. myself, 3 months ago) it's unclear that the IdentityServer extension methods **must** be called instead of ASP.NET Core's own built-in methods. This is further confounded by the fact that IdentityServer4's extension methods are declared in the `Microsoft.AspNetCore.Http` namespace (something I personally disagree with, but nvm) - as visible in this file: https://github.com/IdentityServer/IdentityServer4/blob/e14a1e9b2dfd4173bedb3e570ac5c888927ccb22/src/Extensions/HttpContextAuthenticationExtensions.cs

This also introduces a possible future bug because C#s extension methods are always ranked lower in method overload selection compared to actual type members - so if a future ASP.NET Core update introduces `SignInAsync` methods with matching method signature types it would break IdentityServer4-based projects.

A second, albeit minor, problem exists with the `String name` parameter of IdentityServer4's `SignInAsync` method. The method's XML documentation does not make it clear what `String name` is supposed to refer to: a user's globally-distinct "user name" (login name) claim or their non-distinct "display name" claim - or something else. After some research i saw that IdentityServer4's `SignInAsync` method actually copies `String name` into `IdentityServerUser.DisplayName`, so given that I would expect the method parameter name to be `String displayName` instead. Ideally, source code should abstain from ambiguous public member and parameter names, especially names like `Name` without explicit further context.

**Describe the solution you'd like**

* Rename all of IdentityServer4's `SignInAsync` extension methods to something that makes it explicitly clear it's calling into IdentityServer, e.g. `IdentityServerSignInAsync` or `SignInToIdentityServerAsync`
* And also move the `AuthenticationManagerExtensions` class back into an IdentityServer-owned namespace. I'm of the opinion that extension methods must always be opt-in, and that shared libraries should not declare types in namespaces they don't own (like `Microsoft.AspNetCore.Http`) - by using common non-owned namespaces it makes it impossible to opt-out of using certain extension methods.

* Rename the `String name` parameter of `SignInAsync` to `String displayName` to make it clear which JWT Claim the parameter argument value is being assigned to.

* Audit the project's codebase to rectify the use of any semantically ambiguous identifiers like "name".

* Improve the project's documentation to make it clear the relationship between login-names, user-names, display-names, and JWT Claims like `name`, `preferred_username`, `sub`, `given_name`, etc.

**Describe alternatives you've considered**

* Not applicable - or maybe IdentityServer could add an MSBuild task to verify that IdentityServer's own `SignInAsync` is called at least once instead of ASP.NET Core's method.

**Additional context**

* Not applicable.

**Comment by leastprivilege** [comment]
Thanks - there are some valid points. We will discuss.

**Comment by jamieide** [comment]
"IdentityServer extension methods must be called instead of ASP.NET Core's own built-in methods" Is this true? I have been using the Microsoft.AspNetCore.Authentication extension methods and it seems to be working.

            var scheme = await _authenticationSchemeProvider.GetDefaultAuthenticateSchemeAsync();
            await _httpContextAccessor.HttpContext.SignInAsync(scheme.Name, principal);

I switched over to using the IdSrv extension method and don't notice any difference but I am very new to this. I agree with the suggestions, I am just trying to understand if I am doing something wrong.

**Comment by Jehoel** [comment]
@jamieide I think it's because you're explicitly providing the correct scheme name (`GetDefaultAuthenticateSchemeAsync`) - whereas in my project my default authentication scheme was not the same scheme as IdentityServer scheme name.

**Comment by AndreSteenbergen** [comment]
I just encountered this as well. What could also be an idea, is to provide a signinmanager. Just like asnet identity does. Or simply creating a derived class, as only idp, sub and auth_mr claims are added on the principal. That's what I did in the end. I added identitycore, and added a custom signinmanager, adding idp and auth_mr claims. As sub claim is already added. Just my 2c a slightly different angle.
